### PR TITLE
Convert ChatCompletionMessage to Dict after completion

### PR DESCRIPTION
**Pull request by kevin666aa**
## Why are these changes needed?
1. `extract_text_or_function_call` -> `extract_text_or_completion_object`. The extraction is now a dict or an completion object, changing the name for clarity.

2. Ensure the return from `generate_oai_reply ` is a str or dict, which is a vulnerability.  

Note: Remaining issues with groupchat will be fixed in a separate PR. (When selector is making a function call).

## Related issue number

#721

## Checks

- [x] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [x] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

**Comment by afourney** [comment]
I think this is also related to #765

File: autogen/oai/client.py [hunk 1]
```diff
@@ -343,21 +347,31 @@ def _completions_create(self, client, params):
         return response
 
     @classmethod
-    def extract_text_or_function_call(cls, response: ChatCompletion | Completion) -> List[str]:
-        """Extract the text or function calls from a completion or chat response.
+    def extract_text_or_completion_object(cls, response: ChatCompletion | Completion) -> List[str]:
```
[the remark this answers is not in the document]
Comment (kevin666aa) [reply autogen/oai/client.py]: Yes, revised.

File: autogen/agentchat/conversable_agent.py [hunk 2]
```diff
@@ -625,7 +625,7 @@ def generate_oai_reply(
         response = client.create(
             context=messages[-1].pop("context", None), messages=self._oai_system_message + messages
         )
-        return True, client.extract_text_or_function_call(response)[0]
+        return True, client.extract_text_or_completion_object(response)[0]
```
[the remark this answers is not in the document]
Comment (kevin666aa) [reply autogen/agentchat/conversable_agent.py]: Yes, I think it is related. As commented in conversable_agent: "line 301, line 271 is converting messages to dict. Can be removed after ChatCompletionMessage_to_dict is merged."

The two lines are a temporary solution and is prone to errors. I just added a conversion function to ensure message is dict in `generate_oai_reply`

**Comment by sonichi** [comment]
Does the change start from openai >= 1.1.0 or openai>1.2 ?

File: test/oai/test_client.py [hunk 3]
```diff
@@ -24,7 +30,87 @@ def test_aoai_chat_completion():
...
+                            "location": {"type": "string", "description": "The city and state e.g. San Francisco, CA"},
+                            "unit": {"type": "string", "enum": ["c", "f"]},
+                        },
+                        "required": ["location"],
+                    },
+                },
+            }
+        ],
+    )
+    print(response)
```
Comment (IANTHEREAL) [inline test/oai/test_client.py:69]: What is the purpose of this test, and should we add some validation logic? For instance, to verify that the results returned by the new version api conform to newly logic.

Comment (kevin666aa) [reply test/oai/test_client.py:69]: Just to make sure that `client.extract_text_or_completion_object(response)` works without error.

**Comment by sonichi** [comment]
> > Does the change start from openai >= 1.1.0 or openai>1.2 ?
> 
> From my test, tools can be used after openai >= 1.1.0. Shall we change the version in setup.py?

My impression is that openai 1.2 doesn't cause any error. The error appeared since 1.3. So I thought the breaking change was introduced in 1.3. Could you double check?

**Comment by kevin666aa** [comment]
> > > Does the change start from openai >= 1.1.0 or openai>1.2 ?
> > 
> > 
> > From my test, tools can be used after openai >= 1.1.0. Shall we change the version in setup.py?
> 
> My impression is that openai 1.2 doesn't cause any error. The error appeared since 1.3. So I thought the breaking change was introduced in 1.3. Could you double check?


Both 1.2 and 1.3 is good with function call now. But they also have tool_call which we don't support yet.

File: autogen/oai/client.py [hunk 4]
```diff
@@ -442,21 +446,66 @@ def cost(self, response: Union[ChatCompletion, Completion]) -> float:
...
+        # TODO: in the future openai may support tool call to Code Interpreter, doc retriever, etc.
+        if TOOL_ENABLED and response.tool_calls is not None:
+            dict_msg["tool_calls"] = [
+                {
+                    "type": tool_call.type,
+                    "id": tool_call.id,
+                    "function": dict(tool_call.function),
+                }
+                for tool_call in response.tool_calls
+            ]
```
Comment (sonichi) [inline autogen/oai/client.py:506]: I think there is a simpler method like to_dict_recursive()

Comment (kevin666aa) [reply autogen/oai/client.py:506]:
```
result = openai.ChatCompletion.create(
    messages=messages, model="gpt-3.5-turbo", temperature=0
)
result["choices"][0]["message"].to_dict_recursive()
```
Find some references: https://python.langchain.com/docs/integrations/adapters/openai-old
It is supported in the old one. I also checked the new response, it doesn't offer this feature.

Comment (kevin666aa) [reply autogen/oai/client.py]: Changing to `.model_dump(mode="dict")`

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -49,7 +49,7 @@ jobs:
       - name: Coverage
         if: matrix.python-version == '3.10'
         run: |
-          pip install -e .[mathchat,test]
+          pip install -e .[test]
```
Comment (sonichi) [inline .github/workflows/build.yml:52]: would this reduce the coverage?

File: .github/workflows/openai.yml [hunk 6]
```diff
@@ -4,7 +4,7 @@
 name: OpenAI
 
 on:
-  pull_request_target:
+  pull_request:
```
Comment (sonichi) [inline .github/workflows/openai.yml:7]: please revert this change.

File: .github/workflows/build.yml (hunk 5 above)
Comment (kevin666aa) [reply .github/workflows/build.yml:52]: No, this doesn't affect anything.

File: .github/workflows/openai.yml (hunk 6 above)
Comment (kevin666aa) [reply .github/workflows/openai.yml]: done

**Review by sonichi — APPROVED** [review]
